dali/operators/image/convolution/gaussian_blur.cc

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "dali/core/static_switch.h"
+#include "dali/kernels/imgproc/convolution/separable_convolution_cpu.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/operators/image/convolution/gaussian_blur.h"
+#include "dali/operators/image/convolution/gaussian_blur_params.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/operator/common.h"
+
+namespace dali {
+
+constexpr static const char* kSigmaArgName = "sigma";
+constexpr static const char* kWindowSizeArgName = "window_size";
+
+DALI_SCHEMA(GaussianBlur)
+    .DocStr(R"code(Apply Gaussian Blur to the input.
+Separable convolution with Gaussian kernels is used, proceeding from innermost to outermost
+dimension to calculate the output.
+
+User can specify the sigma or kernel window size.
+If only the sigma is provided, the radius is of kernel is calculated ``ceil(3 * sigma)``,
+thus the kernel window size is ``2 * ceil(3 * sigma) + 1``.
+
+If only the kernel window size is provided, the sigma is calculated using the following formula::
+
+  radius = (window_size - 1) / 2
+  sigma = (radius - 1) * 0.3 + 0.8
+
+Both sigma and kernel window size can be specified as single value for all data axes
+or per data axis.
+
+When specifying the sigma or window size per axis, they are provided same as layouts: from outermost
+to innermost.
+The channel ``C`` and frame ``F`` dimensions are not considered data axes.
+
+For example, with ``HWC`` input, user can provide ``sigma=1.0`` or ``sigma=(1.0, 2.0)`` as there
+are two data axes H and W.
+
+The same input can be provided as per-sample tensors.
+)code")
+    .NumInput(1)
+    .NumOutput(1)
+    .AllowSequences()
+    .SupportVolumetric()
```

Comment:
It would be nice to be able to convert type - at least to "float", meaningful conversion between integers would require a scaling factor.

Reply:
I added support for dtype: unspecified (passes through the input), same as input and float.